### PR TITLE
Initial Mempool sync

### PR DESCRIPTION
**Pull request by sdbondi**
## Description

This PR adds a protocol handler for initial mempool sync called `/tari/mempool-sync/1.0`.

- The protocol handler for the mempool is responsible for the initial sync of transactions from peers.
- In order to prevent duplicate transactions being received from multiple peers, syncing occurs one peer at a time.
- The node will _initiate_ this protocol up to a configurable (`MempoolSyncConfig::num_initial_sync_peers`) number
  of times. After which, it will only respond to sync requests from remote peers.
- After this protocol has completed successfully, both sides should have
  a completely synched mempool.
- Each message has a size limit of 3MiB. The maximum number of initial inventory items is `3MiB / 32 - 1 = 98303`
- This protocol has 1.5 round-trips 

```text
Alice initiates (initiator) the connection to Bob (responder).
As the initiator, Alice MUST send a transaction inventory
Bob SHOULD respond with any transactions known to him, excluding the
transactions in the inventory
Once all transactions have been streamed, Bob MUST send an empty `TransactionItem` (1 byte in protobuf) to indicate that fact to Alice
Bob MAY send indexes relating to Alice's inventory items that are not known to him
Alice SHOULD return the Transactions relating to those indexes
Alice SHOULD close the stream immediately after sending

+-------+                    +-----+
| Alice |                    | Bob |
+-------+                    +-----+
|                                |
| Txn Inventory                  |
|------------------------------->|
|                                |
|      TransactionItem(tx_b1)    |
|<-------------------------------|
|             ...streaming...    |
|      TransactionItem(empty)    |
|<-------------------------------|
|  Inventory missing txn indexes |
|<-------------------------------|
|                                |
| TransactionItem(tx_a1)         |
|------------------------------->|
|             ...streaming...    |
| TransactionItem(empty)         |
|------------------------------->|
|                                |
|             END                |
```

** Network compatibility **

TLDR; No breaking changes. 
A node supporting this protocol will use comms protocol negotiation to determine if a node speaks this protocol. If not, the protocol will fail and the mempool will not be synched (as before).  

**Extras**

- Cleaned up `#[cfg(feature = "base_node")]` by making `cfg_base_node!` macro
- Added tari_comms canonical framing and `open_framed_substream` convenience function

## Motivation and Context

The mempool is not persisted and is empty on node startup until new transactions are propagated on the network. This leads to mining nodes mining empty blocks or newer transactions as they do not have a recent view of transactions in their peer's mempool. This PR  "bootstraps" a node's mempool from peers on startup.

## How Has This Been Tested?

Numerous functional tests.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [x] Documentation

## Checklist:

* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/core/src/mempool/service/initializer.rs [hunk 1]
```diff
@@ -155,22 +163,39 @@ where T: BlockchainBackend + 'static
...
+            mempool_state_event_publisher,
+            self.mempool.clone(),
+            outbound_mp_interface.clone(),
+        );
 
         // Register handle to OutboundMempoolServiceInterface before waiting for handles to be ready
         handles_fut.register(outbound_mp_interface);
         handles_fut.register(local_mp_interface);
 
+        // TODO:
```
Comment (philipr-za) [inline base_layer/core/src/mempool/service/initializer.rs]: The dreaded TODO that can never be done!

File: base_layer/core/src/mempool/sync_protocol/mod.rs [hunk 2]
```diff
@@ -0,0 +1,614 @@
...
+};
+use tari_crypto::{
+    keys::SecretKey,
+    ristretto::RistrettoSecretKey,
+    tari_utilities::{hex::Hex, ByteArray},
+};
+use tokio::{sync::Semaphore, task};
+
+const MAX_FRAME_SIZE: usize = 3 * 1024 * 1024; // 3 MiB
+const LOG_TARGET: &str = "c::mempool::sync-protocol";
```
Comment (philipr-za) [inline base_layer/core/src/mempool/sync_protocol/mod.rs]: Nit/Style question: Should we be sticking to the underscore as a word delimiter?

File: base_layer/core/src/mempool/sync_protocol/mod.rs [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+            "Sending transaction inventory containing {} item(s) to peer `{}`",
+            excess_sigs.len(),
+            self.peer_node_id.short_str()
+        );
+        let inventory = proto::TransactionInventory {
+            items: excess_sigs.iter().map(|sig| sig.to_vec()).collect(),
+        };
+        write_message(framed, inventory).await?;
+
+        read_and_insert_transactions_until_complete(framed, self.mempool.clone(), self.peer_node_id.clone()).await?;
```
Comment (philipr-za) [inline base_layer/core/src/mempool/sync_protocol/mod.rs]: Could have this function as part of the impl with a `&self` argument so that you don't need the mempool and peer_node_id argument?

File: base_layer/core/src/mempool/sync_protocol/mod.rs [hunk 4]
```diff
@@ -0,0 +1,614 @@
...
+            self.peer_node_id.short_str()
+        );
+
+        let transactions = async_mempool::snapshot(self.mempool.clone()).await?;
+        let key_len = <RistrettoSecretKey as SecretKey>::key_length();
+        let excess_sigs = transactions
+            .iter()
+            .map(|txn| txn.body.kernels()[0].excess_sig.get_signature())
+            // Never send a message > the frame size
+            .take(MAX_FRAME_SIZE / key_len)
```
Comment (philipr-za) [inline base_layer/core/src/mempool/sync_protocol/mod.rs]: And the remainder of the transactions? Only going to send maximum the frame size?

File: base_layer/core/src/mempool/sync_protocol/mod.rs [hunk 5]
```diff
@@ -0,0 +1,614 @@
...
+        if num_missing_items > 0 {
+            debug!(target: LOG_TARGET, "Waiting for missing transactions");
+            read_and_insert_transactions_until_complete(framed, self.mempool, self.peer_node_id).await?;
+        }
+
+        Ok(())
+    }
+}
+
+async fn validate_and_insert_transaction<B: BlockchainBackend + 'static>(
```
Comment (philipr-za) [inline base_layer/core/src/mempool/sync_protocol/mod.rs]: This function also seems like it could be in the impl to save some extraneous arguments

**Review by philipr-za — COMMENTED** [review]
Cool to see an implementation of a substream protocol.

File: base_layer/core/src/mempool/service/initializer.rs (hunk 1 above)
Comment (sdbondi) [reply base_layer/core/src/mempool/service/initializer.rs]: `TODO: something better than this **shrugs**`

File: base_layer/core/src/mempool/sync_protocol/mod.rs (hunk 3 above)
Comment (sdbondi) [reply base_layer/core/src/mempool/sync_protocol/mod.rs]: Ah you're right - before I had to consume `self.framed` which meant I couldn't go `self.xxxxx` but at some point I removed the need to do that

Comment (sdbondi) [reply base_layer/core/src/mempool/sync_protocol/mod.rs]: Think there's a few other places I can do that

File: base_layer/core/src/mempool/sync_protocol/mod.rs (hunk 4 above)
Comment (sdbondi) [reply base_layer/core/src/mempool/sync_protocol/mod.rs]: Ye I guess if we've got 98303 transactions in our mempool, we don't need to sync with a peer on startup? Remember this is sending the 32byte kernel excess sig of the transaction so that the receiving node doesn't send transactions that we already know about (i.e exclude txns listed in our inventory) - so if this message is full we _could_ start to get duplicate transactions which is not too bad. I guess a much more reasonable approach may be to just put a limit on how many transactions to sync.

File: base_layer/core/src/mempool/sync_protocol/mod.rs (hunk 2 above)
Comment (sdbondi) [reply base_layer/core/src/mempool/sync_protocol/mod.rs]: Suppose targets refer to the name of the module and so should be `_`

**Review by philipr-za — DISMISSED** [review]
LGTM

**Review by philipr-za — DISMISSED** [review]
LGTM

**Review by philipr-za — DISMISSED** [review]
LGTM

File: base_layer/core/src/lib.rs [hunk 6]
```diff
@@ -31,20 +31,18 @@
...
+
+cfg_base_node! {
+    pub mod blocks;
+    pub mod chain_storage;
+    pub mod consensus;
+    pub mod helpers;
+    pub mod mining;
+    pub mod proof_of_work;
+    pub mod validation;
+}
```
Comment (CjS77) [inline base_layer/core/src/lib.rs]: Does this break intellisense? 
It's the reason we moved away from a previous macro that did this

Comment (sdbondi) [reply base_layer/core/src/lib.rs]: It appears that issue is fixed - no problems with intellisense :)

Comment (sdbondi) [reply base_layer/core/src/lib.rs]: Removed

**Review by CjS77 — APPROVED** [review]
LGTM
